rsocket-examples/src/main/java/io/rsocket/examples/transport/tcp/lease/RateLimitingWithLeaseExample.java

```diff
@@ -0,0 +1,211 @@
+                          }
+                        }))
+            .lease(() -> Leases.create().sender(new LeaseCalculator(SERVER_TAG, messagesQueue)))
+            .bindNow(TcpServerTransport.create("localhost", 7000));
+
+    LeaseReceiver receiver = new LeaseReceiver(CLIENT_TAG);
+    RSocket clientRSocket =
+        RSocketConnector.create()
+            .lease(() -> Leases.create().receiver(receiver))
+            .connect(TcpClientTransport.create(server.address()))
+            .block();
+
+    Objects.requireNonNull(clientRSocket);
+
+    // generate stream of fnfs
+    Flux.generate(
+            () -> 0L,
+            (state, sink) -> {
+              sink.next(state);
+              return state + 1;
+            })
+        // here we wait for the first lease for the responder side and start execution
+        // on if there is allowance
+        .delaySubscription(receiver.notifyWhenNewLease().then())
+        .concatMap(
+            tick -> {
+              System.out.println("Sending " + tick);
+              return Mono.defer(() -> clientRSocket.fireAndForget(ByteBufPayload.create("" + tick)))
+                  .retryWhen(
+                      Retry.indefinitely()
+                          // ensures that error is the result of missed lease
+                          .filter(t -> t instanceof MissingLeaseException)
+                          .doBeforeRetryAsync(
+                              rs -> {
+                                // here we create a mechanism to delay the retry until
+                                // the new lease allowance comes in.
+                                System.out.println("Ran out of leases " + rs);
+                                return receiver.notifyWhenNewLease().then();
+                              }));
+            })
+        .blockLast();
+
+    clientRSocket.onClose().block();
+    server.dispose();
+  }
+
+  /**
+   * This is a class responsible for making decision on whether Responder is ready to receive new
+   * FireAndForget or not base in the number of messages enqueued. <br>
+   * In the nutshell this is responder-side rate-limiter logic which is created for every new
+   * connection.<br>
+   * In real-world projects this class has to issue leases based on real metrics
+   */
+  private static class LeaseCalculator implements Function<Optional<LeaseStats>, Flux<Lease>> {
+    final String tag;
+    final BlockingQueue<?> queue;
+
+    public LeaseCalculator(String tag, BlockingQueue<?> queue) {
+      this.tag = tag;
+      this.queue = queue;
+    }
+
+    @Override
+    public Flux<Lease> apply(Optional<LeaseStats> leaseStats) {
+      System.out.println(
+          String.format("%s stats are %s", tag, leaseStats.isPresent() ? "present" : "absent"));
+      Duration ttlDuration = Duration.ofSeconds(5);
+      // The interval function is used only for the demo purpose and should not be
+      // considered as the way to issue leases.
```

Comment:
It would be helpful to describe what would be used instead of `interval`?

Reply:
Added link to https://github.com/Netflix/concurrency-limits#server-limiter